### ISSUE TITLE
Ichimoku Chikou Span Indicator Calculation Incorrect?

### ISSUE THREAD

**Issue by Andre0711er** [ta4j/ta4j#705]
Hi.

Just evaluating Ichimoku Cloud and i dont know, but might be calculation from Chikou Span is not correct.
If i check your underlying link https://school.stockcharts.com/doku.php?id=technical_indicators:ichimoku_cloud i see the calculation must be 26 Days in the past. So, 26 bars back. In IchimokuChikouSpanIndicator Class i see, calculation is done by "index + timeDelay". So this mean index (now) plus 26 in Front. Is this right?

File: core/indicators/ichimoku/IchimokuChikouSpanIndicator.java
Line: 74

`

protected Num calculate(int index) {
  int spanIndex = index + timeDelay;
  if (spanIndex <= getBarSeries().getEndIndex()) {
    return closePriceIndicator.getValue(spanIndex);
  } else {
    return NaN.NaN;
   }
}
`

**Comment by team172011** [comment]
hi @Andre0711er  thanks for reporting, I will have a closer look

**Comment by MarcDahlem** [comment]
Here is a fix and a complete ichimoku strategy:

https://github.com/MarcDahlem/ta4j/blob/test/test-my-strategies/ta4j-examples/src/test/java/ta4jexamples/strategies/IntelligentTa4jOhlcBenchmarks.java#L442

As trade-dependent indicators https://github.com/ta4j/ta4j/pull/757 , before rules https://github.com/ta4j/ta4j/pull/773 and positive delay indicators https://github.com/ta4j/ta4j/pull/772 are not accepted to be merged here, I cannot make a PR for the corrected Ichimoku indicators and a corresponding example strategy here at ta4j :-/

![grafik](https://user-images.githubusercontent.com/2813850/118689511-6002a500-b807-11eb-9409-400294c6beb0.png)

**Comment by nimo23** [comment]
@MarcDahlem If you need those (so far) not accepted PR's to solve this issue then we should discuss it. If @team172011 is not able to provide a suitable solution then we should accept the solution from @MarcDahlem.

**Comment by team172011** [comment]
@MarcDahlem I am happy to see that there is a possible fix. I remember that there were discussions about this indicator years ago..

We should definitely fix it in ta4j as well. But we should try to follow the established format:

- As far as I know, all the indicators of the _Ichimoku Cloud_ are not depending on the executed trades/positions (like all other indicators). So in general a _trade-dependent_ #757 indicator should not be a requirement for the _Ichimoku Cloud_. 

- Also if there is a need for functionalities like in  _before rule_ #773 or in a _positive delay indicator_ #772 adding this functionality in a scope of a dedicated rule or indicator to this lib is another question.

**Comment by MarcDahlem** [comment]
References to already existing issues/solutions:
* https://github.com/ta4j/ta4j/pull/495
* https://github.com/ta4j/ta4j/issues/448
* https://github.com/ta4j/ta4j/pull/452
* https://github.com/ta4j/ta4j/issues/374
* https://github.com/ta4j/ta4j/issues/614

**Comment by MarcDahlem** [comment]
> 
> 
> @MarcDahlem I am happy to see that there is a possible fix. I remember that there were discussions about this indicator years ago..

Yes I saw that it was often part of discussions. For getting an overview, I linked them all in the previous post.

> We should definitely fix it in ta4j as well. But we should try to follow the established format:
> 
>     * As far as I know, all the indicators of the _Ichimoku Cloud_ are not depending on the executed trades/positions (like all other indicators). So in general a _trade-dependent_ #757 indicator should not be a requirement for the _Ichimoku Cloud_.

Yes and no. The indicators themselves do not need it.

But as soon as you want to use those Ichimoku indicators in (usefull) rules/strategy, you cannot avoid it.
The main point is, that in the most definitions of Ichimoku, the entry conditions (close price did cross cloud  close --> close price > cloud) are still satisfied if the exit criteria (gain reached, or lagging span crossed close price) are met.

In order to avoid this. the indicator must be "reset" (say the crossing up the cloud never happened) as soon as an exit happened.
 
>     * Also if there is a need for functionalities like in  _before rule_ #773 or in a _positive delay indicator_ #772 adding this functionality in a scope of a dedicated rule or indicator to this lib is another question.

The delay indicator is just a helper, not directly needed, but increases readability.

The before rule is needed because:

1. Check if the close price completely crossed the cloud
  * cross of the cloud's lower line will never appear at the same time as the crossing of the upper line. Here, you need cross lower BEFORE cross upper.
  * you cannot simply "hack" it with UnderIndicatorRule( lowestValue(closePrice, X) , cloudLowerLine**AtY**)), becuase you dont know the time Y of the lowest price. And also there, the lowest price in X timeslots does not need to have the lowest of the cloud also at that position
2. Wait for a) the crossing of the laggingSpan up the past cloud + b) crossing of the ask price up the current cloud. They will likely never happen at the same time. Also here, you need crossingCurrent BEFORE crossing laggingSpan. And also here, the LowestPriceIn does not help

**Comment by MarcDahlem** [comment]
I added the corrected ichimoku indicators without the need of the other rules.
I just created another, simpler strategy as example

**Comment by TheCookieLab** [comment]
I'll take another look at this.
